backend/compact-connect/docs/design/README.md

```diff
@@ -31,6 +57,82 @@ efficient processing. A lambda receives messages from the SQS queue. Each messag
+of ability to read and write data, and to administrate users. Read permissions are granted to a user for an entire
+compact or not at all. Data writing and user administration permissions can each be granted to a user per
+compact/jurisdiction combination. All of a compact user's permissions are stored in a DynamoDB record that is associated
+with their own Cognito user id. That record will be used to generate scopes in the Oauth2 token issued to them on login.
+See [Implementation of scopes](#implementation-of-scopes) for a detailed explanation of the design for exactly how
+permissions will be represented by scopes in an access token. See
+[Implementation of permissions](#implementation-of-permissions) for a detailed explanation of the design for exactly
+how permissions are stored and translated into scopes.
+
+#### Compact Executive Directors and Staff
+
+Compact ED level staff can have permission to read all compact data as well as to create and manage users and their
+permissions. They can grant other users the ability to write data for a particular jurisdiction and to create more
+users associated with a particular jurisdiction. They can also delete any user within their compact, so long as that
+user does not have permissions associated with a different compact.
+
+#### Board Executive Directors and Staff
+
+Board ED level staff can have permission to read all compact data, write data to for their own jurisdiction, and to
+create more users that have permissions within their own jurisdiction. They can also delete any user within their
+jurisdiction, so long as that user does not have permissions associated with a different compact or jurisdiction.
+
+#### Implementation of Scopes
+
+AWS Cognito integrates with API Gateway to provide
+[authorizers](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-integrate-with-cognito.html) on an
+API that can verify the tokens issued by a given User Pool and to protect access based on scopes belonging to
+[Resource Servers](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-define-resource-servers.html)
+associated with that User Pool. In the Staff Users user pool, we represent each compact as its own Resource Server, with
+associated scopes. Unfortunately, because resource servers support only up to 100 scopes each, and we would like to
+control permission to write to or administrate each of more than 50 jurisdictions independently, the combinations would
+require more than 100 scopes per resource server.
+
+To design around the 100 scope limit, we will have to split authorization into two layers: coarse- and fine-grained.
+We can rely on the Cognito authorizers to protect our API endpoints based on fewer coarse-grained scopes, then
+protect the more fine-grained access within the API endpoint logic. The Staff User pool resource servers will are
```

Comment:
Nit, maybe silly question: At this point, why not just do all of the access control within the API endpoints? (I would agree this is better, but I wonder what your rationale would be here.)

Reply:
There are a couple of reasons to do it this way:
1) Generally, we want avoid rolling-our-own security as much as is possible. Hand crafting authn/authz logic is intrinsically more risky than relying on dedicated tools that do so for you, so as a general principal, we should defer as much of our authorization to Cognito as we can. Here, we are keeping the most critical pieces in Cognito: Token verification and coarse grained access control. We will only cover fine-grained access controls in our own logic because we can't fit them into Cognito directly.
2) There is a performance benefit to using a dedicated API Gateway authorizer instead of authorizing in runtime: Latency and cost. API Gateway authorizers are very efficient compute-wise in that they sit in front of the compute that we pay for (the endpoint Lambda function) and gate-keep malicious requests such that our compute is never touched. Authorizers can also cache request authorization results so that they don't need to continually be re-validating recently used tokens. Between the gate-keeping and caching, we end up with a much more performant, cost effective API when we use the AWS tools for authorization.

Reply:
Makes a lot of sense. Thanks for the detailed explanation.